### PR TITLE
Allow PdfCopy to be used for writing new pages

### LINKED ISSUE #1165: PdfCopy cannot be used for writing
**Is your feature request related to a problem? Please describe.**
`com.lowagie.text.pdf.PdfCopy` extends `PdfWriter`, but it cannot actually be used for writing to a document. It just silently ignores calls that tries to add new pages.

**Describe the solution you'd like**
`PdfCopy` should just add new pages if requested.

**Describe alternatives you've considered**
An alternative is to generate a temporary PDF file with the required new pages and then copy those into the new document with PdfCopy, but that is a bit wasteful.

**Your real name**
Rasmus Faber-Espensen

**Additional context**
PR will follow shortly.

### PR DESCRIPTION
**Pull request by rasmusfaber**
## Description of the new Feature/Bugfix

Remove unnecessary overridden methods from PdfCopy thus allowing it to be used to write new pages as well as copy them.

Related Issue: This fixes #1165 

## Unit-Tests for the new Feature/Bugfix

- [X] Unit-Tests added to reproduce the bug
- [X] Unit-Tests added to the added feature

## Compatibilities Issues

The behavior of `PdfCopy` is of course changed, since it now actually adds new pages if requested. I would expect that anyone that does that, actually intended to do it.

## Your real name
Rasmus Faber-Espensen

## Testing details

See unit test.

### CODE REVIEW

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfCopy.java [hunk 1]
```diff
@@ -489,17 +489,6 @@ public void close() {
...
-        return null;
-    }
-
-    public void addAnnotation(PdfAnnotation annot) {
-    }
-
-    PdfIndirectReference add(PdfPage page, PdfContents contents) throws PdfException {
-        return null;
-    }
-
```
Comment (asturio) [inline openpdf/src/main/java/com/lowagie/text/pdf/PdfCopy.java:502]: ok, so addAnnotation(PdfAnnotation) and add(PdfPage, PdfContent) are not overwriting the methods of PdfWriter anymore, and so the methods of PdfWriter would be called.
I see that add(PdfOutline) has no parent method. And deleting if could be ok, as the method doesn't do anything.

I'm just not sure, if allowing PdfCopy to manipulate the content of the document is what the original author planed for the class.

**Comment by asturio** [comment]
This change will have twofold results:

- Semantics of PdfCopy will change
- PdyCopy can be used to merge PDFs.

Leaving the PR open for the community to drop some 2 cents.

**Comment by mkl-public** [comment]
> Leaving the PR open for the community to drop some 2 cents.

You mentioned in review:

> I'm just not sure, if allowing PdfCopy to manipulate the content of the document is what the original author planed for the class.

Not only have they not planned that, this feature has not been tested for years in use. I would not be surprised if there were numerous corner cases for which the feature will fail.

Thus, a larger number of tests of different situations would be nice. We have first adding content, then merging; other tests could be first merging, then adding content, probably also a longer mix of mergers and content additions.

Nonetheless, the feature of course is interesting, it also was missing in all pre-7 versions of iText and there were requests again and again.

**Comment by rasmusfaber** [comment]
> Thus, a larger number of tests of different situations would be nice. We have first adding content, then merging; other tests could be first merging, then adding content, probably also a longer mix of mergers and content additions.

The single test I wrote already added/merged/added. I just added another test that interleaves two files and combines it with adding a page in front of each.

**Comment by rasmusfaber** [comment]
> > I'm just not sure, if allowing PdfCopy to manipulate the content of the document is what the original author planed for the class.
> 
> Not only have they not planned that, this feature has not been tested for years in use. I would not be surprised if there were numerous corner cases for which the feature will fail.

Just for reference, none of the other `PdfWriter` specializations (`PdfStamper`, `PdfCopyFields`, `PdfCopyForms` etc) override these methods. If there are corner cases where this don't work, they most likely don't work for those classes either - and at least the code tries to do something now instead of just silently doing nothing.

File: openpdf/src/main/java/com/lowagie/text/pdf/PdfCopy.java (hunk 1 above)
Comment (rasmusfaber) [reply openpdf/src/main/java/com/lowagie/text/pdf/PdfCopy.java:502]: > I'm just not sure, if allowing PdfCopy to manipulate the content of the document is what the original author planed for the class.

You can already do the exact same thing:

```
PdfCopy copy = new PdfSmartCopy(document, outputStream)) {
document.open();
document.add(new Paragraph("Front page"));
document.newPage();
```
should do the exact same thing as the much more verbose and complicated
```
PdfCopy copy = new PdfSmartCopy(document, outputStream)) {
document.open();

ByteArrayOutputStream baos = new ByteArrayOutputStream();
Document tempDocument = new Document();
PdfWriter tempWriter = PdfWriter.getInstance(tempDocument, baos);
tempDocument.open();
tempDocument.add(new Phrase("Front page"));
tempDocument.close();
PdfReader tempReader = new PdfReader(baos.toByteArray());
PdfImportedPage page = copy.getImportedPage(tempReader, 1);
copy.addPage(page);
copy.freeReader(tempReader);
```

**Comment by mkl-public** [comment]
> I just added another test that interleaves two files and combines it with adding a page in front of each.

Great, thanks!

> Just for reference, none of the other `PdfWriter` specializations (`PdfStamper`, `PdfCopyFields`, `PdfCopyForms` etc) override these methods.

Neither of those classes _is_ a `PdfWriter` specializations. They all _wrap_ `PdfWriter` specializations (`PdfStamperImp`, `PdfCopyFieldsImp`, `PdfCopyFormsImp`). Admittedly, it is easy to access the respective specialization using `getWriter`.

> If there are corner cases where this don't work, they most likely don't work for those classes either - and at least the code tries to do something now instead of just silently doing nothing.

Maybe one should give this feature in `PdfCopy` a chance to prove itself.
